# Math Riddle: Johnny the (Old Fashioned) Painter

<b>Johnny has been hired to paint the numbers 1 through 100.</b><br>
<i> Johnny is old fashioned because he paints numbers below the power of 2 <b>and</b> because we'll be solving using a flat list iterator with low generalizability.</i><br>
<b>   How many times will he paint the letter "2"?</b><br>

<br>
<center> <h3> Johnny!</h3> </center>
<br>
<img style="-webkit-user-select: none;cursor: zoom-in;" src="https://image1.masterfile.com/getImage/846-08140099em-1930s-1940s-man-in-white-overalls-by-step-ladder-zinc-pail-bucket.jpg" width="360" height="283">
<br>
<center> <i> He's a great painter and has a simple yet fullfilling life.</i> </center>
<br>

<b>We'll solve using list comprehensions!</b>

<li>1. Create a list of numbers through 100.</li>
<li>2. Flatten that list to strings of each digit.</li>
<li>3. Define a function that iterates over your list and counts the number</li>
<li>4. Ask for user input and push out a result!</li>

In [1]:
# Step 1 – Create a list of numbers

z=[]
for i in range(100):
    z.append(str(i))
    
# Print your stuff:
print('Your list "z" has a length of: {0}\n'.format(len(z)))
print(z)

####################################

# Step 2 – Flatten that list.

z_flat = [y for x in z for y in x]

# Print your stuff:
print('\n\nYour flat list "z_flat" has a length of: {0}\n'.format(len(z_flat)))
print(z_flat)

####################################

# Step 3 – Define a function that iterates and counts

def item_counter(string_to_look_for = '2', list_to_look_in = z_flat):
    item_count = 0
    for item in list_to_look_in:
        if item == string_to_look_for:
            item_count +=1
    return item_count

# Print your stuff:
print('\n\nYour function "item counter" has been added successfully.')

def item_counter_2(string_to_look_for = '2', list_to_look_in = z):
    item_count = 0    
    for item in list_to_look_in:
        if string_to_look_for in item:
            item_count +=1
    return item_count
item_counter_2()

# Print your stuff:
print('\n\nYour function "item counter" has been added successfully.')


Your list "z" has a length of: 100

['0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12', '13', '14', '15', '16', '17', '18', '19', '20', '21', '22', '23', '24', '25', '26', '27', '28', '29', '30', '31', '32', '33', '34', '35', '36', '37', '38', '39', '40', '41', '42', '43', '44', '45', '46', '47', '48', '49', '50', '51', '52', '53', '54', '55', '56', '57', '58', '59', '60', '61', '62', '63', '64', '65', '66', '67', '68', '69', '70', '71', '72', '73', '74', '75', '76', '77', '78', '79', '80', '81', '82', '83', '84', '85', '86', '87', '88', '89', '90', '91', '92', '93', '94', '95', '96', '97', '98', '99']


Your flat list "z_flat" has a length of: 190

['0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '1', '0', '1', '1', '1', '2', '1', '3', '1', '4', '1', '5', '1', '6', '1', '7', '1', '8', '1', '9', '2', '0', '2', '1', '2', '2', '2', '3', '2', '4', '2', '5', '2', '6', '2', '7', '2', '8', '2', '9', '3', '0', '3', '1', '3', '2', '3', '3', '3', '4', '3', '5', '3', '6', '3',

In [2]:
# Step 4 – Ask for user input and execute

num_look_for = input('Choose a number you want to count in John\'s painting job:\n')
if int(num_look_for) >9:
    list_use = z
else:
    list_use = z_flat
    
    
####################################

# Print your stuff: (The results)

print('\nJohnny paints {0} a total of {1} times.\n'.format(num_look_for,
                                                       item_counter(num_look_for,list_use)))

Choose a number you want to count in John's painting job:
4

Johnny paints 4 a total of 20 times.



# How about a faster way to do it:

In [3]:
# Step 4 – Ask for user input and execute

num_look_for = input('Choose a number you want to count in Johnny\'s painting job:\n')
if int(num_look_for) >9:
    count_value = 3
else:
    count_value = ''.join(str(i) for i in range(1,101)).count(num_look_for)
    
    
    
####################################

# Print your stuff: (The results)

print('\nJohnny paints {0} a total of {1} times.\n'.format(num_look_for,count_value))


Choose a number you want to count in Johnny's painting job:
4

Johnny paints 4 a total of 20 times.



# Janie
<b>Janie has been hired to paint the numbers 1 through 1,000,000.</b><br>
    <i> She's got a very hard job, especially compared to Johnny from the good old days. But Janie loves what she does. Except, she for counting how many times she'll paint a certain number! This part of the job frsutrates her to no end. Let's help her?</i><br>
   <b>How many times will she paint "19"?</b><br>

<hr>

<img style="-webkit-user-select: none;cursor: zoom-in;" src="https://media.istockphoto.com/photos/paintercollection-in-my-gallery-picture-id524298601?k=6&amp;m=524298601&amp;s=612x612&amp;w=0&amp;h=8ltm2aOOynE9nZ1_o_LmLqUSyaojt0DKlfVXbPIrTXI=" width="360" height="456">
<br>
<center> <i> She's been painting for a while now... <br>
    It's not the paint which wears her down... it's the counting.<br>
    Fortunately, our programming skills are going to save her! </i> </center>
<br>

## Create the function

In [4]:
def fast_item_counter(string_to_look_for = '19',
                      list_to_look_in = range(1,1000001)):
    item_count = 0 
    # for every number in the list:
    for i in list_to_look_in:
        # if '19' is in that nunber:
        if string_to_look_for in str(i):
            # add to my count...
            item_count +=1 
            # Take the numbers after the instance of 19
            # For example 1192 becomes 2)
            item_split = str(i).split(string_to_look_for,1)[1]
            # While '19' is in the split string:
            while string_to_look_for in item_split: 
                print(i) # print that number...
                print(item_split) # print the split too...
                if string_to_look_for in item_split:
                    # add to my count
                    item_count +=1
                    # Then split again!
                    item_split = item_split.split(string_to_look_for,1)[1]

          
    return item_count
fast_item_counter()

1919
19
11919
19
19019
019
19119
119
19190
190
19191
191
19192
192
19193
193
19194
194
19195
195
19196
196
19197
197
19198
198
19199
199
19219
219
19319
319
19419
419
19519
519
19619
619
19719
719
19819
819
19919
919
21919
19
31919
19
41919
19
51919
19
61919
19
71919
19
81919
19
91919
19
101919
19
111919
19
119019
019
119119
119
119190
190
119191
191
119192
192
119193
193
119194
194
119195
195
119196
196
119197
197
119198
198
119199
199
119219
219
119319
319
119419
419
119519
519
119619
619
119719
719
119819
819
119919
919
121919
19
131919
19
141919
19
151919
19
161919
19
171919
19
181919
19
190019
0019
190119
0119
190190
0190
190191
0191
190192
0192
190193
0193
190194
0194
190195
0195
190196
0196
190197
0197
190198
0198
190199
0199
190219
0219
190319
0319
190419
0419
190519
0519
190619
0619
190719
0719
190819
0819
190919
0919
191019
1019
191119
1119
191190
1190
191191
1191
191192
1192
191193
1193
191194
1194
191195
1195
191196
1196
191197
1197
191198
1198
191199
1199
191219
1219
19131

50000

<b><i>My numbers get split accordingly:</b></i> <br><br>
**1919:** 19<br>
**11919:** 19 <br>
**19019:** 019 <br>
**19119:** 119 <br>
**19190:** 190 <br>
**19191:** 191 <br>
**19192:** 192

## We want to test out our function:

We'll cycle through the two numbers: 191,918 and 191,919.

* Our function should split the first number into: 1918 <br>
* Our function should split the second number into: 1919 and then 19<br>
* **Our total count should be: 5**


In [5]:
z = fast_item_counter('19',range(191918,191920))
print('\nThe number "19" was found a total of {0} times'.format(z))

191918
1918
191919
1919
191919
19

The number "19" was found a total of 5 times


Indeed, the function does as it should:
* 191,918 was split once and counted twice
* 191,919 was split twice and counted three times
* **Total result was 5!**



## Now that the testing is over, remove any "print" fields unnecessary to core execution of the function:

In [6]:
def fast_item_counter(string_to_look_for = 19,
                      min_val = 0,
                      max_val = 100001):
    
    if type(string_to_look_for) != str:
        string_to_look_for = str(string_to_look_for)
    
    # Create that range!
    range_look = range(min_val,max_val)

    item_count = 0 
    # for every number in the list:
    for i in range_look:
        # if '19' is in that nunber:
        if string_to_look_for in str(i):
            # add to my count...
            item_count +=1 
            # Take the numbers after the instance of 19
            # For example 1192 becomes 2)
            item_split = str(i).split(string_to_look_for,1)[1]
            # While '19' is in the split string:
            while string_to_look_for in item_split: 
                # No printing, only good hard results
                if string_to_look_for in item_split:
                    # add to my count
                    item_count +=1
                    # Then split again!
                    item_split = item_split.split(string_to_look_for,1)[1]           
    return item_count


In [7]:
num_look_for = input('Choose a number you want to count in Janie\'s painting job:\n')
#range_to_look_in = [int(x) for x in input('Choose the range you want to use:\n').split(',')]
# range(range_to_look_in[0],range_to_look_in[1]
min_val = 0
max_val = 100001
count_value = fast_item_counter(num_look_for,min_val,max_val)
    
    
####################################

# Print your stuff: (The results)
print('\nJanie paints {0} a total of {1} times.\nWhen she paints {2} to {3}.'.format(num_look_for,count_value,min_val, max_val))


Choose a number you want to count in Janie's painting job:
10

Janie paints 10 a total of 4001 times.
When she paints 0 to 100001.


# Make Interactive!

## Add your interactive elements

In [8]:
import ipywidgets as widgets
#from ipywidgets import interact,interactive,fixed,interact_manual
import time

#from ipywidgets import IntSlider, Accordion,Button,Checkbox,FloatRangeSlider,IntProgress,BoundedIntText, Dropdown
from IPython.display import display

In [9]:
min_slider = widgets.IntSlider(
    value = 0,
    min = 0,
    max = 9999,
    description = 'First Number: ',
    # use stuff below
    #style = {'handle_color': 'blue',
    #        'alpha': '0.2'},
    disabled = False,
    continuous_update = False,
    readout = True, # default
    readout_format = 'd'
)

min_slider.style.handle_color = 'lightblue'

def on_val_change(change):
    time.sleep(0.1)
    #print('New Value is: {0}\n..................'.format(change['new']))

min_slider.observe(on_val_change,
               names = 'value')

display(min_slider)

IntSlider(value=0, continuous_update=False, description='First Number: ', max=9999, style=SliderStyle(handle_c…

In [10]:
max_slider = widgets.IntSlider(
    value = 500000,
    min = 1,
    max = 999999,
    description = 'Last Number: ',
    # use stuff below
    #style = {'handle_color': 'blue',
    #        'alpha': '0.2'},
    disabled = False,
    continuous_update = False,
    readout = True, # default
    readout_format = 'd'
)

max_slider.style.handle_color = 'darkblue'

max_slider.observe(on_val_change,
               names = 'value')

display(max_slider)

IntSlider(value=500000, continuous_update=False, description='Last Number: ', max=999999, min=1, style=SliderS…

In [11]:
# Add a range slider
range_slider = widgets.FloatRangeSlider(
    value = [0,10000],
    min = 0,
    max = 10000,
    step = 1,
    description = 'Your Range',
    disabled = False,
    continuous_update = False,
    readout = True, # default
    readout_format = 'd'
)

def on_val_change(change):
    time.sleep(0.1)
    #print('New value: {}'.format(change['new']))

range_slider.observe(on_val_change,
               names = 'value')

display(range_slider)

FloatRangeSlider(value=(0.0, 10000.0), continuous_update=False, description='Your Range', max=10000.0, readout…

In [12]:
text_input = widgets.BoundedIntText(
    value = str(2),
    min = 0,
    max = 9999999,
    step = 1,
    description = 'Which number do you want to count?',
    style = {'description_width':'initial'},
    disabled = False
)


text_input.observe(on_val_change,
               names = 'value')
display(text_input)

BoundedIntText(value=2, description='Which number do you want to count?', max=9999999, style=DescriptionStyle(…

In [49]:
min_val_input = widgets.BoundedIntText(
    value = 1,
    min = 0,
    max = 9999999,
    step = 1,
    description = 'What is the first number you paint?',
    style = {'description_width':'initial'},
    disabled = False)

min_val_input.observe(on_val_change,
               names = 'value')

display(min_val_input)

BoundedIntText(value=1, description='What is the first number you paint?', max=9999999, style=DescriptionStyle…

In [52]:
max_val_input = widgets.BoundedIntText(
    value = 10001,
    min = 0,
    max = 9999999,
    step = 1,
    description = 'What is the last number you paint?',
    style = {'description_width':'initial'},
    disabled = False
)

max_val_input.observe(on_val_change,
               names = 'value')
display(max_val_input)

BoundedIntText(value=10001, description='What is the last number you paint?', max=9999999, style=DescriptionSt…

In [97]:
reset = widgets.ToggleButton(
        value=False,
        description='Clear Output',
        disabled=False,
        button_style='warning', # 'success', 'info', 'warning', 'danger' or ''
        tooltip='Clear Output',
        icon='magic')
display(reset)

ToggleButton(value=False, button_style='warning', description='Clear Output', icon='magic', tooltip='Clear Out…

## Add to interactive display

In [54]:
w = widgets.interact(fast_item_counter,
             string_to_look_for=text_input,
             min_val = min_val_input,
             max_val = max_val_input,
         continuous_update=True)

#print(w.widget.result)


interactive(children=(BoundedIntText(value=18, description='Which number do you want to count?', max=9999999, …

In [55]:
# Need to add this to the output thing!
print('\nJanie paints \'{0}\' a total of {1} times\nWhen she paints {2} to {3}.'.format(
    w.widget.kwargs['string_to_look_for'],
    w.widget.result,
    w.widget.kwargs['min_val'],
    w.widget.kwargs['max_val']))


Janie paints '18' a total of 300 times
When she paints 1 to 10001.


In [100]:
out = widgets.Output(layout={'border':'1px solid black'})
#out.clear_ouput()
with out:
    print('... Your Output ...')
    display(w.widget)

# Alternatively...

In [56]:
def new_fast_item_counter(
    string_to_look_for=5,
    min_val = 1,
    max_val = 100000):
    x = fast_item_counter(string_to_look_for,min_val,max_val)
    return print ('Janie paints \'{0}\' a total of {1} times\nWhen she paints {2} to {3}.'.format(
        string_to_look_for, x, min_val,max_val))
    

In [57]:
new_fast_item_counter(51)

Janie paints '51' a total of 4000 times
When she paints 1 to 100000.


In [58]:
w = widgets.interact(new_fast_item_counter,
             string_to_look_for=text_input,
             min_val = min_val_input,
             max_val = max_val_input,
         continuous_update=True)

interactive(children=(BoundedIntText(value=18, description='Which number do you want to count?', max=9999999, …

In [101]:
def on_button_clicked(b):
    if out:
        out.clear_output()
        with out:
            print("Ouput reset.")
            if w.widget:
                display(w.widget)

reset.observe(on_button_clicked)    

# Final Product!

In [103]:
display(reset)
display(out)

ToggleButton(value=True, button_style='warning', description='Clear Output', icon='magic', tooltip='Clear Outp…

Output(layout=Layout(border='1px solid black'), outputs=({'output_type': 'stream', 'text': 'Ouput reset.\n', '…

# Interactiv as output

In [112]:
from ipywidgets import IntSlider
from ipywidgets.embed import embed_minimal_html
slider = IntSlider(value=40)
embed_minimal_html('export.html', views=[slider], title='Widgets export')

In [113]:
embed_minimal_html('export 2.html', views=[out], title='Widgets export 2')